tools/nimdigger.nim

```diff
@@ -0,0 +1,327 @@
+https://stackoverflow.com/a/22592593/1426932 (Magic exit statuses)
+> anything above 127 makes the bisection fail with something like:
+> 125 is magic and makes the run be skipped with git bisect skip.
+]#
+
+import std/[os, osproc, strformat, macros, strutils, tables, algorithm]
+
+proc `$`(a: ref): string =
+  if a == nil: "nil" else: $a[]
+
+template dbg(args: varargs[untyped]): untyped =
+  # so users can swap in their own better logging until stdlib has one
+  echo args
+
+type
+  DiggerOpt = object ## nimdigger input
+    rev: string
+    nimDir: string
+    compileNim: bool
+    fetch: bool
+    csourcesBuildArgs: string
+    buildAllCsources: bool
+    verbose: bool
+
+    # bisect cmds
+    # TODO: allow user to not compile nim, for cases where it's not needed
+    oldnew: string # eg: v0.20.0~10..v0.20.0
+    bisectCmd: string # eg: bin/nim c --hints:off --skipparentcfg --skipusercfg $timn_D/tests/nim/all/t12329.nim 'arg1 bar' 'arg2'
+    bisectBugfix: bool
+  CsourcesOpt = ref object
+    url: string
+    dir: string
+    rev: string
+    binDir: string
+    csourcesBuildArgs: string
+    revs: seq[string]
+    fetch: bool
+    name: string
+    nimCsourcesExe: string
+  DiggerState = ref object ## nimdigger internal state
+    coptv0, coptv1: CsourcesOpt
+    binDir: string
+    nimDir: string
+    rev: string
+
+const
+  csourcesRevs = "v0.9.4 v0.13.0 v0.15.2 v0.16.0 v0.17.0 v0.17.2 v0.18.0 v0.19.0 v0.20.0 64e34778fa7e114b4afc753c7845dee250584167".split
+  csourcesV1Revs = "a8a5241f9475099c823cfe1a5e0ca4022ac201ff".split
+  NimDiggerEnv = "NIMDIGGER_HOME"
+  ExeExt2 = when ExeExt.len > 0: "." & ExeExt else: ""
+
+var verbose = false
+
+proc isSimulate(): bool =
+  defined(nimDiggerSimulate)
+
+proc runCmd(cmd: string) =
+  # TODO: allow `dir` param (or use `runCmdOutput`)
+  if isSimulate():
+    dbg cmd
+  else:
+    if verbose: dbg cmd
+    doAssert execShellCmd(cmd) == 0, cmd
+
+proc runCmdOutput(cmd: string, dir = ""): string =
+  if verbose: dbg cmd, dir
+  let (outp, status) = execCmdEx(cmd, workingDir = dir)
+  doAssert status == 0, indent(&"status: {status}\ncmd: {cmd}\ndir: {dir}\noutput: {outp}", 2)
+  result = outp
+  stripLineEnd(result)
+
+macro ctor(obj: untyped, a: varargs[untyped]): untyped =
+  ## Generates an object constructor call from a list of fields.
+  # xxx expose in std/sugar, factor with https://github.com/nim-lang/fusion/pull/32
+  runnableExamples:
+    type Foo = object
+      a, b: int
+    doAssert Foo.ctor(a,b) == Foo(a: a, b: b)
+  result = nnkObjConstr.newTree(obj)
+  for ai in a: result.add nnkExprColonExpr.newTree(ai, ai)
+
+proc parseKeyVal(a: string): OrderedTable[string, string] =
+  ## parse bash-like entries of the form key=val
+  let a2 = a.splitLines
+  for i, ai in a2:
+    if ai.len == 0 or ai.startsWith "#": continue
+    let b = split(ai, "=", maxsplit = 1)
+    doAssert b.len == 2, $(ai, b)
+    result[b[0]] = b[1]
```

Comment:
```suggestion
proc parseKeyVal(s: string): OrderedTable[string, string] =
  ## parse bash-like entries of the form key=val
  for index, line in splitLines(s):
    if line.len == 0 or line.startsWith "#":
      continue

    let item = split(line, "=", maxsplit = 1)
    doAssert item.len == 2, $(line, item)

    result[item[0]] = item[1]
```

Reply:
this doesn't compile, try it; that's because of iterator vs proc overload

Reply:
I've now simplified this code